### PR TITLE
Support nonlinear positivity limiting for arbitrary variables

### PR DESCRIPTION
**Pull request by bennibolm**
Although you were able to set the limiting variable with `positivity_variables_nonlinear=(pressure,)`, the limiting was hard-coded due to
```
newton_loops_alpha!(alpha, var_min[i, j, element], u_local, i, j, element,
                    pressure_goal, pressure_dgoal_dbeta,
                    pressure_initialCheck, pressure_finalCheck,
                    dt, mesh, equations, dg, cache, limiter)
```
Now, it is implemented for general variables (if the specific functions for initial/final check and the derivative of the variable `variable(u, equations, True())` have been defined):
```
newton_loops_alpha!(alpha, var_min[i, j, element], u_local, i, j, element,
                    variable, initial_check,
                    final_check,
                    dt, mesh, equations, dg, cache, limiter)
```

### CODE REVIEW

File: src/solvers/dgsem_tree/subcell_limiters_2d.jl [hunk 1]
```diff
@@ -554,27 +554,25 @@ end
...
 
     return nothing
 end
 
-pressure_goal(bound, u, equations) = bound - pressure(u, equations)
-function pressure_dgoal_dbeta(u, dt, antidiffusive_flux, equations)
-    -dot(dpdu(u, equations), dt * antidiffusive_flux)
+goal_function(variable, bound, u, equations) = bound - variable(u, equations)
+function dgoal_function(variable, u, dt, antidiffusive_flux, equations)
+    -dot(variable(u, equations, true), dt * antidiffusive_flux)
```
Comment (amrueda) [inline src/solvers/dgsem_tree/subcell_limiters_2d.jl]:
```suggestion
    -dot(variable(u, equations, True()), dt * antidiffusive_flux)
```

Comment (bennibolm) [reply src/solvers/dgsem_tree/subcell_limiters_2d.jl]: Thanks

File: src/equations/compressible_euler_2d.jl [hunk 2]
```diff
@@ -1581,6 +1583,9 @@ end
 
     return (equations.gamma - 1.0) * SVector(0.5 * v_square, -v1, -v2, 1.0)
 end
+@inline function pressure(u, equations::CompressibleEulerEquations2D, derivative::False)
+    return pressure(u, equations)
+end
```
Comment (bennibolm) [inline src/equations/compressible_euler_2d.jl:1588]: The idea of using this third parameter is to just pass one variable to the nonlinear positivity limiter and get both, the variable itself and the derivative. There might be a better way to implement this.

```suggestion
```
I added this just for completeness but then, I should add it for both entropys as well. In fact, I think I will delete this since we will never call `variable(u, equations, False())`. What do you think?

File: src/equations/compressible_euler_2d.jl [hunk 3]
```diff
@@ -1570,6 +1571,7 @@ end
 
     return SVector(w1, w2, w3, w4)
 end
+entropy_spec(u, equations, derivative::True) = cons2entropy_spec(u, equations)
```
Comment (bennibolm) [inline src/equations/compressible_euler_2d.jl]: We never call `entropy_spec(u, equations, False())`. Actually, this isn't even define. Is there a nicer way to use the same function name for the variable itself and its derivative?

Also for `entropy_math` and `pressure`.

Comment (amrueda) [reply src/equations/compressible_euler_2d.jl]: I don't know if there is a nicer way. I would leave it like this for now and postpone this discussion for the merge to Trixi main.

File: test/test_structured_2d.jl [hunk 4]
```diff
@@ -187,8 +187,8 @@ isdir(outdir) && rm(outdir, recursive=true)
 
   @trixi_testset "elixir_euler_source_terms_sc_subcell.jl" begin
     @test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_euler_source_terms_sc_subcell.jl"),
-      l2   = [0.008160127272557726, 0.008658253869683077, 0.009351900401871649, 0.02775701488343099],
```
Comment (amrueda) [inline test/test_structured_2d.jl:190]: @bennibolm Why did you need to change the expected results?

**Review by amrueda — COMMENTED** [review]
LGTM!
I leave a single comment in this review.

File: test/test_structured_2d.jl (hunk 4 above)
Comment (bennibolm) [reply test/test_structured_2d.jl:190]: After I changed `cons2entropy_spec` to `entropy_spec(.., .., True())`, I got slightly different errors. Later, I found out that this was because of the missing `@inline`. Somehow, this seems to have some very small rounding differences. Then, I added it again in [this](https://github.com/bennibolm/Trixi.jl/pull/115/commits/463de3abccbe605612758e8b11254f3a36c72a74) commit.
For some possible performance increase, I added `@inline` to the check routines in the [next](https://github.com/bennibolm/Trixi.jl/pull/115/commits/e4d3a9d6b573908769db0e6c2ed6614a46d3e1d4) commit as well. It looks like the change is so small that all other tests still work, but some (simulations with non-Cartesian meshes seem to be more sensitive for changes) change a bit.

Comment (bennibolm) [reply test/test_structured_2d.jl:190]: But in general, all tests were created a long time ago, and it might be possible that small rounding difference came across in the meanwhile, e.g. through the change in the time integration method with nominator and denominator.
Over the time, I had to fix some tests, but not all. So, the actual errors might slightly differ from the given ones, but only in a margin such that the tests do not fail.
